SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityOfferHolder.java

```diff
@@ -25,25 +26,31 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(SingularityMesosScheduler.class);
 
-  private final Protos.Offer offer;
+  private final List<Protos.Offer> offers;
   private final List<SingularityTask> acceptedTasks;
   private final Set<SingularityPendingTaskId> rejectedPendingTaskIds;
   private List<Resource> currentResources;
+  private Set<String> roles;
 
   private final String rackId;
+  private final String slaveId;
+  private final String hostname;
   private final String sanitizedHost;
   private final String sanitizedRackId;
 
   private final Map<String, String> textAttributes;
   private final Map<String, String> reservedSlaveAttributes;
 
-  public SingularityOfferHolder(Protos.Offer offer, int taskSizeHint, String rackId, Map<String, String> textAttributes, Map<String, String> reservedSlaveAttributes) {
+  public SingularityOfferHolder(List<Protos.Offer> offers, int taskSizeHint, String rackId, String slaveId, String hostname, Map<String, String> textAttributes, Map<String, String> reservedSlaveAttributes) {
     this.rackId = rackId;
-    this.offer = offer;
-    this.acceptedTasks = Lists.newArrayListWithCapacity(taskSizeHint);
-    this.currentResources = offer.getResourcesList();
+    this.slaveId = slaveId;
+    this.hostname = hostname;
+    this.offers = offers;
+    this.roles = MesosUtils.getRoles(offers.get(0));
+    this.acceptedTasks = Lists.newArrayListWithExpectedSize(taskSizeHint);
+    this.currentResources = offers.size()  > 1 ? MesosUtils.combineResources(offers.stream().map(Protos.Offer::getResourcesList).collect(Collectors.toList())) : offers.get(0).getResourcesList();
```

Comment:
I think a more general solution might be to sort out which offers to accept/decline/cache when launching the tasks at the end. i.e. if we have an offers of 2,3 and 4 cpus, and matched tasks that use 3 and 3, we can hold on to the 3 cpu offer, and use the other two. No reason to calculate that as we go along since the total resources is the same, but we can certainly do it at the end with the goal of maximizing usage of the current offers.

#lets-do-the-math